code/modules/atmospherics/machinery/pipes/smart.dm

```diff
@@ -0,0 +1,115 @@
+/obj/machinery/atmospherics/pipe/smart
+	icon = 'icons/obj/atmospherics/pipes/simple.dmi'
+	icon_state = "pipe11-3"
+
+	name = "pipe"
+	desc = "A one meter section of regular pipe."
+
+	vis_flags = VIS_INHERIT_ICON | VIS_INHERIT_ICON_STATE | VIS_INHERIT_DIR | VIS_INHERIT_ID
+
+	device_type = QUATERNARY
+	construction_type = /obj/item/pipe/quaternary
+	pipe_state = "manifold4w"
+	var/connection_num = 0
+	var/list/connections
+	var/static/list/mutable_appearance/center_cache = list()
+	var/mutable_appearance/pipe_appearance
+
+/* We use New() instead of Initialize() because these values are used in update_icon()
+ * in the mapping subsystem init before Initialize() is called in the atoms subsystem init.
+ */
+/obj/machinery/atmospherics/pipe/smart/New()
+	icon_state = ""
+	connections = new/list(dir2text(NORTH) = FALSE, dir2text(SOUTH) = FALSE , dir2text(EAST) = FALSE , dir2text(WEST) = FALSE)
+	return ..()
+
+/obj/machinery/atmospherics/pipe/smart/Initialize()
+	. = ..()
+	update_appearance()
+
+/obj/machinery/atmospherics/pipe/smart/SetInitDirections()
+	initialize_directions = ALL_CARDINALS
+
+/obj/machinery/atmospherics/pipe/smart/proc/check_connections()
+	var/mutable_appearance/center
+	connection_num = 0
+	connections = list(dir2text(NORTH) = FALSE, dir2text(SOUTH) = FALSE , dir2text(EAST) = FALSE , dir2text(WEST) = FALSE)
+	var/static/list/valid_connectors = typecacheof(/obj/machinery/atmospherics)
+	for(var/direction in connections)
+		var/turf/turf = get_step(src, text2dir(direction))
+		if(!turf)
+			return
+		for(var/machine_type in turf.contents)
```

Comment:
this is fucking with my head, why are you using a typecache for this?

Reply:
to be able to quickly search... what i should use instead?

Reply:
i changed the system, now it will use just a for loop with a typecheck